### PR TITLE
Kill all subprocesses properly on windows

### PR DESCRIPTION
**Pull request by steelbrain**
Fixes https://github.com/steelbrain/exec/issues/80

### CODE REVIEW

**Comment by steelbrain** [comment]
Will investigate the failure on AppVeyor on a windows PC soon

File: src/index.js [hunk 1]
```diff
@@ -95,9 +96,8 @@ async function exec(
 
     if (options.timeout !== Infinity) {
       timeout = setTimeout(function() {
-        try {
-          spawnedProcess.kill()
-        } catch (_) { /* No Op */ }
+        // eslint-disable-next-line no-use-before-define
```
Comment (Arcanemagus) [inline src/index.js]: Why not just define it before use...?

File: src/index.js [hunk 2]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
 }
 
+function killProcess(spawnedProcess: Object, signal: string = 'SIGTERM'): void {
+  const spawnfile = Path.basename(spawnedProcess.spawnfile)
+  if (process.platform !== 'win32' && spawnfile !== 'wmic.exe' && spawnfile !== 'wmic') {
+    // Also do this if the process is wmic
+    spawnedProcess.kill(signal)
+    return
+  }
+  execProxy('wmic', ['process', 'where', `(ParentProcessId=${spawnedProcess.pid})`, 'get', 'processid'], { stream: 'stdout', timeout: 60 * 1000 })
```
Comment (Arcanemagus) [inline src/index.js]: Can this be split over multiple lines? Or the arguments / options brought out into constants?

File: src/index.js [hunk 3]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
+    .then(function(output: any) {
```
Comment (Arcanemagus) [inline src/index.js]: Any reason not to make this function `async` and `await` it above?

File: src/index.js [hunk 4]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
+      const pids = output.split(/\s+/).filter(i => /^\d+$/.test(i)).map(parseInt).filter(i => i !== spawnedProcess.pid && i > 0)
+      pids.forEach(function(pid) {
+        process.kill(pid, signal)
+      })
+    }).catch(function(error) {
+      console.error('[sb-exec] Error killing process on windows', error)
```
Comment (Arcanemagus) [inline src/index.js]: "windows" -> "Windows"

File: src/index.js [hunk 5]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
+    spawnedProcess.kill(signal)
+    return
+  }
+  execProxy('wmic', ['process', 'where', `(ParentProcessId=${spawnedProcess.pid})`, 'get', 'processid'], { stream: 'stdout', timeout: 60 * 1000 })
+    .then(function(output: any) {
+      const pids = output.split(/\s+/).filter(i => /^\d+$/.test(i)).map(parseInt).filter(i => i !== spawnedProcess.pid && i > 0)
+      pids.forEach(function(pid) {
+        process.kill(pid, signal)
+      })
+    }).catch(function(error) {
```
Comment (Arcanemagus) [inline src/index.js]: The Atom version of this code specifically ignores errors from the `wmic` process, should this work in the same manner?

File: src/index.js [hunk 6]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
       killSignal = givenKillSignal || 'SIGTERM'
     }
   }
   return promise
 }
 
+function killProcess(spawnedProcess: Object, signal: string = 'SIGTERM'): void {
+  const spawnfile = Path.basename(spawnedProcess.spawnfile)
+  if (process.platform !== 'win32' && spawnfile !== 'wmic.exe' && spawnfile !== 'wmic') {
+    // Also do this if the process is wmic
```
Comment (Arcanemagus) [inline src/index.js]: This comment doesn't make any sense, since the `if` specifically excludes all calls to `wmic`.

File: src/index.js [hunk 7]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
     } else {
       killSignal = givenKillSignal || 'SIGTERM'
     }
   }
   return promise
 }
 
+function killProcess(spawnedProcess: Object, signal: string = 'SIGTERM'): void {
+  const spawnfile = Path.basename(spawnedProcess.spawnfile)
+  if (process.platform !== 'win32' && spawnfile !== 'wmic.exe' && spawnfile !== 'wmic') {
```
Comment (Arcanemagus) [inline src/index.js]: This logic seems off, the first conditional excludes Windows platforms, but the second conditional can only ever be valid on Windows. Also, if the platform isn't Windows it should be using this method regardless of whether the `spawnFile` is `wmic` or not.

File: src/index.js [hunk 8]
```diff
@@ -114,21 +114,41 @@ function execProxy(filePath: string, parameters: Array<string> = [], givenOption
...
+    // Also do this if the process is wmic
+    spawnedProcess.kill(signal)
+    return
+  }
+  execProxy('wmic', ['process', 'where', `(ParentProcessId=${spawnedProcess.pid})`, 'get', 'processid'], { stream: 'stdout', timeout: 60 * 1000 })
+    .then(function(output: any) {
+      const pids = output.split(/\s+/).filter(i => /^\d+$/.test(i)).map(parseInt).filter(i => i !== spawnedProcess.pid && i > 0)
+      pids.forEach(function(pid) {
+        process.kill(pid, signal)
```
Comment (Arcanemagus) [inline src/index.js]: Atom ignores errors in attempting to kill these child processes, while here we are propagating them up to the `catch` on L147.

**Comment by steelbrain** [comment]
Beta build on AppVeyor is having Atom issues, stable passed so merging in

**Pull request merged**